In [18]:
# Fatemeh Makhloughi
# Machine Learning
# HW9 - Decision Tree and Random Forest  classifier for digits dataset classification
from sklearn.datasets import load_digits
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics


In [2]:
digits = load_digits()

In [3]:
x = digits.data
y = digits.target

In [4]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [5]:
clf = DecisionTreeClassifier(random_state=0)

In [6]:
clf.fit(x_train, y_train)
y_pred = clf.predict(x_test)

In [7]:
print("Accuracy: %0.2f" % clf.score(x_test, y_test))

Accuracy: 0.85


In [8]:
print(confusion_matrix(y_test, y_pred))

[[29  0  1  0  2  1  0  0  0  0]
 [ 0 22  1  0  1  0  0  1  2  1]
 [ 1  1 28  2  0  0  1  0  0  0]
 [ 0  0  0 30  1  0  0  1  1  1]
 [ 0  0  0  0 40  2  1  1  0  2]
 [ 0  0  1  1  1 41  2  0  0  1]
 [ 0  0  0  0  1  0 33  0  1  0]
 [ 0  0  1  1  1  1  0 29  0  1]
 [ 0  2  1  2  1  1  0  2 19  2]
 [ 0  0  0  3  1  1  0  1  0 34]]


In [11]:
rf_model = RandomForestClassifier(n_estimators=50, max_features="auto", random_state=44)

In [19]:
 rf_model.fit(x_train, y_train)

RandomForestClassifier(n_estimators=50, random_state=44)

In [22]:
y_pred = rf_model.predict(x_test)

In [23]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9805555555555555


In [24]:
print(confusion_matrix(y_test, y_pred))

[[33  0  0  0  0  0  0  0  0  0]
 [ 0 28  0  0  0  0  0  0  0  0]
 [ 0  0 33  0  0  0  0  0  0  0]
 [ 0  0  0 33  0  1  0  0  0  0]
 [ 0  0  0  0 46  0  0  0  0  0]
 [ 0  0  0  0  0 46  1  0  0  0]
 [ 0  0  0  0  0  1 34  0  0  0]
 [ 0  0  0  0  0  0  0 33  0  1]
 [ 0  1  0  0  0  0  0  0 29  0]
 [ 0  0  0  0  0  1  0  1  0 38]]


The results were obtained using the random forest method better than the decision tree. Accuracy for decision tree is 85% while for random forest it is 98%.